### PR TITLE
Key/value backend

### PR DESCRIPTION
**Pull request by moretea**
There are quite some ideas in the issue tracker about how we could improve nixops radically.

I would like to start out with a feasible iterative plan. As far as I can see, only very trivial SQL operations are every executed against the sqlite3 database; the main use case is to use transactions to support rollback in the case of exceptions. This should not be to hard to implement ourselves in memory, in combination with a key/value server that supports locking certain keys.

- [x] Refactor the statefile and deployment lock into a specific local implementation
- [x] Introduce different URI schema's for future backends
- [ ] Refactor all the sqlite3 queries into concrete operations implemented in `file.Statefile`
    - [x] Deployment
    - [x] Resource
    - [x] VirtualBox
    - [ ] GCE _done, needs to be verified_
    - [ ] AWS _done, needs to be verified_
    - [ ] Hetzner _done, needs to be verified_
- [x] Implement a local json file backend

### CODE REVIEW

File: nixops/state/__init__.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    scheme = url.scheme
+
+    if scheme == "":
+        scheme = "file"
+
+    def raise_(ex):
+        raise ex
+
+    switcher = {
+            "file": lambda(url): file.StateFile(url.path),
```
Comment (bjornfor) [inline nixops/state/__init__.py]: Nitpick: wrong indentation

**Comment by bjornfor** [comment]
I'm not a nixops developer, but IMHO this looks nice.

**Comment by moretea** [comment]
Update: So now this works OK with running the  virtualbox tests! (except for this one https://github.com/NixOS/nixops/issues/622)

**Comment by moretea** [comment]
And that last commit fixed the atomic transactions in the implementation of resources.

**Comment by kamilchm** [comment]
Do we really need to have 2 different ways to manage state - `atomic`/`not atomic`?
It looks more complicated to move this decision to backends instead of state implementation.

**Comment by moretea** [comment]
@kamilchm I believe that you're right. I thought that I needed that for the json backend, but in the end that was not necessary at all.

**Comment by moretea** [comment]
I removed the following two items from the TODO list; I believe that they should be tacked in another PR.
Actually, the first item is part of the JSON file backend that I already created in this PR.

- [x] Add a general 'JSON' based backend (that has a deployment key => document, where the document contains both the deployment attributes, all resources and resource attributes).
- [ ] Implement state and locking backed by e.g. `etcd`

**Comment by moretea** [comment]
@rbvermaa @kamilchm are there things that I can do to make it easier to get this merged :)?

File: doc/manual/hacking.xml [hunk 2]
```diff
@@ -49,10 +49,20 @@ linkend="sec-deploying-to-ec2"/>.)  To run a specific test, run
...
 <screen>
 $ python2 tests.py tests.functional.test_encrypted_links
 </screen>
 
+To filter on which backends you want to run functional tests against, you can
+filter on one or more tags.
+To run e.g. only the virtualbox tests, run:
+
+<screen>
+$ python 2 tests.py tests.functional -A vbox
```
Comment (rbvermaa) [inline doc/manual/hacking.xml]: Remove space between python and 2.

**Comment by rbvermaa** [comment]
@edolstra Do you have any comments to this PR?

**Comment by grahamc** [comment]
@moretea I replaced my nixops install with this via:

```
    # pkgs.nixops
    (import ./nixops/release.nix {}).build.x86_64-linux
```

and got:

```
======================================================================
ERROR: Failure: ImportError (No module named file)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/nix/store/367q5apg6bapf79rwmmm7kha57sji81l-python2.7-nose-1.3.7/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/nix/store/367q5apg6bapf79rwmmm7kha57sji81l-python2.7-nose-1.3.7/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/nix/store/367q5apg6bapf79rwmmm7kha57sji81l-python2.7-nose-1.3.7/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/tmp/nix-build-nixops-1.5.1pre0_abcdef.drv-0/nixops-1.5.1pre0_abcdef/tests/__init__.py", line 6, in <module>
    import nixops.state.file
ImportError: No module named file
```

**Comment by moretea** [comment]
Yes, just found that out myself when trying to run the tests one final time; the issue was that I invoked the tests in another way that did not trigger this :(. The commit that I pushed after your comment should fix that, I'm running the vbox tests again at the moment.

**Comment by grahamc** [comment]
Ok I pulled again and re-tested:

```
$ nixops deploy --include builder-19 
Traceback (most recent call last):
  File "/nix/store/59qbcipcy3hrykqbxdbiwdzpj8vpxrq8-nixops-1.5.1pre0_abcdef/bin/..nixops-wrapped-wrapped", line 9, in <module>
    import nixops.state.file
ImportError: No module named state.file
```

**Comment by moretea** [comment]
@grahamc that should fix it!

**Comment by grahamc** [comment]
Not sure:

```
$ nixops deploy --include builder-19 
Traceback (most recent call last):
  File "/nix/store/xdl2ql51sclf1yr5xi3cab2qgx03mi0s-nixops-1.5.1pre0_abcdef/bin/..nixops-wrapped-wrapped", line 9, in <module>
    import nixops.state.sqlite3_file
ImportError: No module named state.sqlite3_file
```

python's import system is a gift that keeps on giving :D

**Comment by ip1981** [comment]
Wish: don't store in state file parameters that can be extracted from nix expressions or not used anyway: e. g. EC2 instance type, keys, `toplevel`, `storeKeysOnMachine`, `configsPath`, etc.

**Comment by nh2** [comment]
What's the motiviation behind this motion? Is it to be able to support other backends than sqlite for the purpose of having the state stored e.g. in a networked database?

**Comment by moretea** [comment]
Exactly.

Op 21 jun. 2017 00:35 schreef "Niklas Hambüchen" <notifications@github.com>:

**Comment by rbvermaa** [comment]
Thanks, committed in 308ac51556

**Comment by domenkozar** [comment]
@rbvermaa wrong issue?

**Comment by bachp** [comment]
I would like to test this for AWS, what's the easiest way to get this version in my setup?

**Comment by bachp** [comment]
@moretea @grahamc I got as far that I now also getting the error

```
Traceback (most recent call last):
  File "/nix/store/czy2gz21dp72i6lfm1k2gx98ss8kl51x-nixops-1.6pre0_abcdef/bin/..nixops-wrapped-wrapped", line 9, in <module>
    import nixops.state.sqlite3_file
ImportError: No module named state.sqlite3_file
```

Have you been able to solve this?

**Comment by mogorman** [comment]
fixed merge conflicts in my clone of the fork and will continue modifying the source for this patch there.

**Comment by andrewchambers** [comment]
Some explanation of the reason/point of this change would be nice. The original post provides no rationale.

**Comment by moretea** [comment]
The point is that you'll be able to store state remotely eventually, instead of a local sqlite3 file.

Being able to have a key/value that stores to JSON will be a first step to be able to store settings in a _remote_ implementation of that

**Comment by comozo** [comment]
Hi,
@moretea this feature would be great for us. Is that possible to merge it ?
Ta.

**Comment by domenkozar** [comment]
See https://github.com/NixOS/nixops/pull/1264

**Comment by grahamc** [comment]
Hello!

Thank you for this PR. 

In the past several months, some major changes have taken place in
NixOps:

1. Backends have been removed, preferring **a plugin-based architecture**.
   Here are some of them:
   
   - https://github.com/NixOS/nixops-aws
   - https://github.com/NixOS/nixops-hetzner
   - https://github.com/nix-community/nixops-vbox
   - https://github.com/nix-community/nixops-libvirtd
   - https://github.com/nix-community/nixops-datadog
   
2. NixOps Core has been updated to be **Python 3 only**, and at the
   same time, **MyPy type hints** have been added and are now strictly
   required during CI.

This is all accumulating in to what I hope will be a **NixOps 2.0
release**. There is a tracking issue for that:
https://github.com/NixOS/nixops/issues/1242 . It is possible that 
more core changes will be made to NixOps for this release, with a
focus on simplifying NixOps core and making it easier to use and work
on.

My hope is that by adding types and more thorough automated testing,
it will be **easier for contributors to make improvements**, and for
contributions like this one to merge in the future.

However, because of the major changes, it has become likely that this
PR cannot merge right now as it is. The backlog of now-unmergable PRs
makes it hard to see which ones are being kept up to date.

If you would like to see this merge, **please bring it up to date with
`master` and reopen it**. If the or mypy type checking fails, please
correct any issues and then reopen it. I will be looking primarily at
open PRs whose tests are all green. 

Thank you again for the work you've done here, I am sorry to be
closing it now.

Graham

**Pull request closed without merging**